File: day01.py
from itertools import combinations
from math import prod

def read_puzzle(filename):
    with open(filename) as f:
         return list(map(int, f.readlines()))

def ex1( puzzle ):
    for item in combinations(puzzle, 2):
        if sum(item) == 2020:
            return prod(item)

def ex2( puzzle ):
    for item in combinations(puzzle, 3):
        if sum(item) == 2020: 
            return prod(item)

puzzle = read_puzzle("input01.txt")
print(ex1(puzzle))
print(ex2(puzzle))

File: AdventOfCode2020/day01.jl
import Base.Iterators: product

function read_puzzle(filename)

    return open(filename) do f
        return parse.(Int, readlines(f))
    end

end

function ex1(puzzle)
    for item in product(puzzle, puzzle)
        if length(unique(item)) == 2
            sum(item) == 2020  && return prod(item)
        end
    end
    return
end

function ex2(puzzle)
    for item in product(puzzle, puzzle, puzzle)
        if length(unique(item)) == 3
            sum(item) == 2020 && return prod(item)
        end
    end
    return
end

puzzle = read_puzzle("input01.txt")
println(ex1(puzzle))
println(ex2(puzzle))
